torchdata/dataloader2/utils/worker.py

```diff
@@ -0,0 +1,121 @@
+# LICENSE file in the root directory of this source tree.
+
+import random
+
+from dataclasses import dataclass
+from typing import Callable, Optional
+
+import torch
+
+from torchdata.dataloader2.graph import DataPipe
+from torchdata.dataloader2.utils import generate_random_int
+from torchdata.datapipes.iter import IterDataPipe
+from torchdata.datapipes.map import MapDataPipe
+
+try:
+    import numpy
+
+    HAS_NUMPY = True
+except ModuleNotFoundError:
+    HAS_NUMPY = False
+
+
+@dataclass(frozen=True)
+class DistInfo:
+    r"""
+    Message class for keeping track of distributed information.
+
+    Args:
+        world_size (int): Total number of distributed nodes
+        rank (int): Distributed rank for the current distributed node
+    """
+    world_size: int = 1
+    rank: int = 0
+
+
+@dataclass(frozen=True)
+class WorkerInfo:
+    r"""
+    Message class for keeping track of worker information.
+
+    Args:
+        num_workers (int): Total number of worker processes
+        worker_id (int): Worker ID for the current worker process
+    """
+    num_workers: int
+    worker_id: int
+
+
+@dataclass(frozen=True)
+class _ExtraInfo:
+    r"""
+    Message class for extra arguments.
+    """
+    shared_seed: int
+
+
+def process_init_fn(
+    datapipe: DataPipe,
+    dist_info: DistInfo,
+    worker_info: WorkerInfo,
+    custom_init_fn: Optional[Callable[[DataPipe, DistInfo, WorkerInfo], DataPipe]] = None,
+) -> DataPipe:
+    r"""
+    Based on the distributed and worker information, shard the ``DataPipe`` graph dynamically.
+    """
+    global_worker_id = worker_info.worker_id * dist_info.world_size + dist_info.rank
+    total_num_workers = worker_info.num_workers * dist_info.world_size
+    torch.utils.data.graph_settings.apply_sharding(datapipe, total_num_workers, global_worker_id)
+
+    if custom_init_fn is not None:
+        datapipe = custom_init_fn(datapipe, dist_info, worker_info)
+        assert isinstance(datapipe, (IterDataPipe, MapDataPipe))
+
+    return datapipe
+
+
+def process_reset_fn(
+    datapipe: DataPipe,
+    dist_info: DistInfo,
+    worker_info: WorkerInfo,
+    extra_info: _ExtraInfo,
+    custom_reset_fn: Optional[Callable[[DataPipe, DistInfo, WorkerInfo], DataPipe]] = None,
+) -> DataPipe:
+    r"""
+    Based on the distributed shared random seed and worker id, this function is used to
+    reset the random state of the ``DataPipe`` graph and the global random states for ``torch``,
+    ``random`` and ``numpy``.
+    """
+    # This function will receive worker local copy of datapipe and reset function from ``initialize_iteration``
+    worker_seed_generator = torch.Generator()
+    worker_seed_generator.manual_seed(extra_info.shared_seed)
+    torch.utils.data.graph_settings.apply_random_seed(
+        datapipe,
+        worker_seed_generator,
+    )
+    # Set different seeds across distributed workers
+    global_worker_id = worker_info.worker_id * dist_info.world_size + dist_info.rank
```

Comment:
If I understand correctly, Say we have 2 distributed training process (world_size=2) and 3 data loading subprocess per training process (num_workers=3). the global_worker_id assignment is like following:

```
rank        0   1
------------------
worker0     0   1     
worker1     2   3
worker2     4   5
```

Essentially the global worker id with rank 0 are (0, 2, 4) while the global worker id with rank 1 are (1, 3, 5). Wondering why not make global worker id with the same rank continuous? ~

Reply:
> Essentially the global worker id with rank 0 are (0, 2, 4) while the global worker id with rank 1 are (1, 3, 5). Wondering why not make global worker id with the same rank continuous? ~

Aha, you found a tricky problem we found earlier. We can't make the same rank continuous because we want to make sure data are sharded as evenly as possible. Using the same example with `world_size=2, num_workers=3`, if the total number of data is 8 and we do continuous global worker id per rank, the length of data in rank 0 will be 5 but rank 1 will be 3. And, if I make the `global_worker_id` as the current implementation, both ranks will be able to fetch 4 data

Reply:
@ejguan  I see. Because the sharding mechanism is something like `tuple_id % total_worker == global_worker_id`? So when it's not divisible, lower "global_worker_id" will get the "extra one row". Result in more rows in low rank worker?~

Reply:
I am not 100% sure that I understand.
We will try to fill in low workers for all nodes from low rank to high rank.